// @tanstack/angular-table/build/lib/esm2022/flex-render.mjs
import { ChangeDetectorRef, Directive, effect, Inject, inject, Injector, Input, runInInjectionContext, TemplateRef, ViewContainerRef, } from '@angular/core';
import { FlexRenderComponentProps } from './flex-render/context';
import { FlexRenderFlags } from './flex-render/flags';
import { flexRenderComponent, } from './flex-render/flex-render-component';
import { FlexRenderComponentFactory } from './flex-render/flex-render-component-ref';
import { FlexRenderComponentView, FlexRenderTemplateView, mapToFlexRenderTypedContent, } from './flex-render/view';
import { memo } from '@tanstack/table-core';
import * as i0 from "@angular/core";
export { injectFlexRenderContext, } from './flex-render/context';
export class FlexRenderDirective {
    viewContainerRef;
    templateRef;
    #flexRenderComponentFactory = inject(FlexRenderComponentFactory);
    #changeDetectorRef = inject(ChangeDetectorRef);
    content = undefined;
    props = {};
    injector = inject(Injector);
    renderFlags = FlexRenderFlags.ViewFirstRender;
    renderView = null;
    #latestContent = () => {
        const { content, props } = this;
        return typeof content !== 'function'
            ? content
            : runInInjectionContext(this.injector, () => content(props));
    };
    #getContentValue = memo(() => [this.#latestContent(), this.props, this.content], latestContent => {
        return mapToFlexRenderTypedContent(latestContent);
    }, { key: 'flexRenderContentValue', debug: () => false });
    constructor(viewContainerRef, templateRef) {
        this.viewContainerRef = viewContainerRef;
        this.templateRef = templateRef;
    }
    ngOnChanges(changes) {
        if (changes['props']) {
            this.renderFlags |= FlexRenderFlags.PropsReferenceChanged;
        }
        if (changes['content']) {
            this.renderFlags |=
                FlexRenderFlags.ContentChanged | FlexRenderFlags.ViewFirstRender;
            this.update();
        }
    }
    ngDoCheck() {
        if (this.renderFlags & FlexRenderFlags.ViewFirstRender) {
            // On the initial render, the view is created during the `ngOnChanges` hook.
            // Since `ngDoCheck` is called immediately afterward, there's no need to check for changes in this phase.
            this.renderFlags &= ~FlexRenderFlags.ViewFirstRender;
            return;
        }
        this.renderFlags |= FlexRenderFlags.DirtyCheck;
        const latestContent = this.#getContentValue();
        if (latestContent.kind === 'null' || !this.renderView) {
            this.renderFlags |= FlexRenderFlags.ContentChanged;
        }
        else {
            this.renderView.content = latestContent;
            const { kind: previousKind } = this.renderView.previousContent;
            if (latestContent.kind !== previousKind) {
                this.renderFlags |= FlexRenderFlags.ContentChanged;
            }
        }
        this.update();
    }
    update() {
        if (this.renderFlags &
            (FlexRenderFlags.ContentChanged | FlexRenderFlags.ViewFirstRender)) {
            this.render();
            return;
        }
        if (this.renderFlags & FlexRenderFlags.PropsReferenceChanged) {
            if (this.renderView)
                this.renderView.updateProps(this.props);
            this.renderFlags &= ~FlexRenderFlags.PropsReferenceChanged;
        }
        if (this.renderFlags &
            (FlexRenderFlags.DirtyCheck | FlexRenderFlags.DirtySignal)) {
            if (this.renderView)
                this.renderView.dirtyCheck();
            this.renderFlags &= ~(FlexRenderFlags.DirtyCheck | FlexRenderFlags.DirtySignal);
        }
    }
    #currentEffectRef = null;
    render() {
        if (this.#shouldRecreateEntireView() && this.#currentEffectRef) {
            this.#currentEffectRef.destroy();
            this.#currentEffectRef = null;
            this.renderFlags &= ~FlexRenderFlags.RenderEffectChecked;
        }
        this.viewContainerRef.clear();
        this.renderFlags =
            FlexRenderFlags.Pristine |
                (this.renderFlags & FlexRenderFlags.ViewFirstRender) |
                (this.renderFlags & FlexRenderFlags.RenderEffectChecked);
        const resolvedContent = this.#getContentValue();
        if (resolvedContent.kind === 'null') {
            this.renderView = null;
        }
        else {
            this.renderView = this.#renderViewByContent(resolvedContent);
        }
        // If the content is a function `content(props)`, we initialize an effect
        // in order to react to changes if the given definition use signals.
        if (!this.#currentEffectRef && typeof this.content === 'function') {
            this.#currentEffectRef = effect(() => {
                this.#latestContent();
                if (!(this.renderFlags & FlexRenderFlags.RenderEffectChecked)) {
                    this.renderFlags |= FlexRenderFlags.RenderEffectChecked;
                    return;
                }
                this.renderFlags |= FlexRenderFlags.DirtySignal;
                // This will mark the view as changed,
                // so we'll try to check for updates into ngDoCheck
                this.#changeDetectorRef.markForCheck();
            }, { injector: this.viewContainerRef.injector });
        }
    }
    #shouldRecreateEntireView() {
        return (this.renderFlags &
            FlexRenderFlags.ContentChanged &
            FlexRenderFlags.ViewFirstRender);
    }
    #renderViewByContent(content) {
        if (content.kind === 'primitive') {
            return this.#renderStringContent(content);
        }
        else if (content.kind === 'templateRef') {
            return this.#renderTemplateRefContent(content);
        }
        else if (content.kind === 'flexRenderComponent') {
            return this.#renderComponent(content);
        }
        else if (content.kind === 'component') {
            return this.#renderCustomComponent(content);
        }
        else {
            return null;
        }
    }
    #renderStringContent(template) {
        const context = () => {
            return typeof this.content === 'string' ||
                typeof this.content === 'number'
                ? this.content
                : this.content?.(this.props);
        };
        const ref = this.viewContainerRef.createEmbeddedView(this.templateRef, {
            get $implicit() {
                return context();
            },
        });
        return new FlexRenderTemplateView(template, ref);
    }
    #renderTemplateRefContent(template) {
        const latestContext = () => this.props;
        const view = this.viewContainerRef.createEmbeddedView(template.content, {
            get $implicit() {
                return latestContext();
            },
        });
        return new FlexRenderTemplateView(template, view);
    }
    #renderComponent(flexRenderComponent) {
        const { inputs, outputs, injector } = flexRenderComponent.content;
        const getContext = () => this.props;
        const proxy = new Proxy(this.props, {
            get: (_, key) => getContext()[key],
        });
        const componentInjector = Injector.create({
            parent: injector ?? this.injector,
            providers: [{ provide: FlexRenderComponentProps, useValue: proxy }],
        });
        const view = this.#flexRenderComponentFactory.createComponent(flexRenderComponent.content, componentInjector);
        return new FlexRenderComponentView(flexRenderComponent, view);
    }
    #renderCustomComponent(component) {
        const view = this.#flexRenderComponentFactory.createComponent(flexRenderComponent(component.content, {
            inputs: this.props,
            injector: this.injector,
        }), this.injector);
        return new FlexRenderComponentView(component, view);
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.11", ngImport: i0, type: FlexRenderDirective, deps: [{ token: ViewContainerRef }, { token: TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.11", type: FlexRenderDirective, isStandalone: true, selector: "[flexRender]", inputs: { content: ["flexRender", "content"], props: ["flexRenderProps", "props"], injector: ["flexRenderInjector", "injector"] }, providers: [FlexRenderComponentFactory], usesOnChanges: true, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.11", ngImport: i0, type: FlexRenderDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[flexRender]',
                    standalone: true,
                    providers: [FlexRenderComponentFactory],
                }]
        }], ctorParameters: () => [{ type: i0.ViewContainerRef, decorators: [{
                    type: Inject,
                    args: [ViewContainerRef]
                }] }, { type: i0.TemplateRef, decorators: [{
                    type: Inject,
                    args: [TemplateRef]
                }] }], propDecorators: { content: [{
                type: Input,
                args: [{ required: true, alias: 'flexRender' }]
            }], props: [{
                type: Input,
                args: [{ required: true, alias: 'flexRenderProps' }]
            }], injector: [{
                type: Input,
                args: [{ required: false, alias: 'flexRenderInjector' }]
            }] } });
