// @tanstack/angular-table/build/lib/esm2022/flex-render/flags.mjs
/**
 * Flags used to manage and optimize the rendering lifecycle of content of the cell,
 * while using FlexRenderDirective.
 */
export var FlexRenderFlags;
(function (FlexRenderFlags) {
    /**
     * Indicates that the view is being created for the first time or will be cleared during the next update phase.
     * This is the initial state and will transition after the first ngDoCheck.
     */
    FlexRenderFlags[FlexRenderFlags["ViewFirstRender"] = 1] = "ViewFirstRender";
    /**
     * Represents a state where the view is not dirty, meaning no changes require rendering updates.
     */
    FlexRenderFlags[FlexRenderFlags["Pristine"] = 2] = "Pristine";
    /**
     * Indicates the `content` property has been modified or the view requires a complete re-render.
     * When this flag is enabled, the view will be cleared and recreated from scratch.
     */
    FlexRenderFlags[FlexRenderFlags["ContentChanged"] = 4] = "ContentChanged";
    /**
     * Indicates that the `props` property reference has changed.
     * When this flag is enabled, the view context is updated based on the type of the content.
     *
     * For Component view, inputs will be updated and view will be marked as dirty.
     * For TemplateRef and primitive values, view will be marked as dirty
     */
    FlexRenderFlags[FlexRenderFlags["PropsReferenceChanged"] = 8] = "PropsReferenceChanged";
    /**
     * Indicates that the current rendered view needs to be checked for changes.
     */
    FlexRenderFlags[FlexRenderFlags["DirtyCheck"] = 16] = "DirtyCheck";
    /**
     * Indicates that a signal within the `content(props)` result has changed
     */
    FlexRenderFlags[FlexRenderFlags["DirtySignal"] = 32] = "DirtySignal";
    /**
     * Indicates that the first render effect has been checked at least one time.
     */
    FlexRenderFlags[FlexRenderFlags["RenderEffectChecked"] = 64] = "RenderEffectChecked";
})(FlexRenderFlags || (FlexRenderFlags = {}));
