// @tanstack/angular-table/build/lib/esm2022/flex-render/flex-render-component-ref.mjs
import { ChangeDetectorRef, inject, Injectable, KeyValueDiffers, OutputEmitterRef, ViewContainerRef, } from '@angular/core';
import * as i0 from "@angular/core";
export class FlexRenderComponentFactory {
    #viewContainerRef = inject(ViewContainerRef);
    createComponent(flexRenderComponent, componentInjector) {
        const componentRef = this.#viewContainerRef.createComponent(flexRenderComponent.component, {
            injector: componentInjector,
        });
        const view = new FlexRenderComponentRef(componentRef, flexRenderComponent, componentInjector);
        const { inputs, outputs } = flexRenderComponent;
        if (inputs)
            view.setInputs(inputs);
        if (outputs)
            view.setOutputs(outputs);
        return view;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.11", ngImport: i0, type: FlexRenderComponentFactory, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
    static ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.11", ngImport: i0, type: FlexRenderComponentFactory });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.11", ngImport: i0, type: FlexRenderComponentFactory, decorators: [{
            type: Injectable
        }] });
export class FlexRenderComponentRef {
    componentRef;
    componentInjector;
    #keyValueDiffersFactory;
    #componentData;
    #inputValueDiffer;
    #outputRegistry;
    constructor(componentRef, componentData, componentInjector) {
        this.componentRef = componentRef;
        this.componentInjector = componentInjector;
        this.#componentData = componentData;
        this.#keyValueDiffersFactory = componentInjector.get(KeyValueDiffers);
        this.#outputRegistry = new FlexRenderComponentOutputManager(this.#keyValueDiffersFactory, this.outputs);
        this.#inputValueDiffer = this.#keyValueDiffersFactory
            .find(this.inputs)
            .create();
        this.#inputValueDiffer.diff(this.inputs);
        this.componentRef.onDestroy(() => this.#outputRegistry.unsubscribeAll());
    }
    get component() {
        return this.#componentData.component;
    }
    get inputs() {
        return this.#componentData.inputs ?? {};
    }
    get outputs() {
        return this.#componentData.outputs ?? {};
    }
    /**
     * Get component input and output diff by the given item
     */
    diff(item) {
        return {
            inputDiff: this.#inputValueDiffer.diff(item.inputs ?? {}),
            outputDiff: this.#outputRegistry.diff(item.outputs ?? {}),
        };
    }
    /**
     *
     * @param compare Whether the current ref component instance is the same as the given one
     */
    eqType(compare) {
        return compare.component === this.component;
    }
    /**
     * Tries to update current component refs input by the new given content component.
     */
    update(content) {
        const eq = this.eqType(content);
        if (!eq)
            return;
        const { inputDiff, outputDiff } = this.diff(content);
        if (inputDiff) {
            inputDiff.forEachAddedItem(item => this.setInput(item.key, item.currentValue));
            inputDiff.forEachChangedItem(item => this.setInput(item.key, item.currentValue));
            inputDiff.forEachRemovedItem(item => this.setInput(item.key, undefined));
        }
        if (outputDiff) {
            outputDiff.forEachAddedItem(item => {
                this.setOutput(item.key, item.currentValue);
            });
            outputDiff.forEachChangedItem(item => {
                if (item.currentValue) {
                    this.#outputRegistry.setListener(item.key, item.currentValue);
                }
                else {
                    this.#outputRegistry.unsubscribe(item.key);
                }
            });
            outputDiff.forEachRemovedItem(item => {
                this.#outputRegistry.unsubscribe(item.key);
            });
        }
        this.#componentData = content;
    }
    markAsDirty() {
        this.componentRef.injector.get(ChangeDetectorRef).markForCheck();
    }
    setInputs(inputs) {
        for (const prop in inputs) {
            this.setInput(prop, inputs[prop]);
        }
    }
    setInput(key, value) {
        if (this.#componentData.allowedInputNames.includes(key)) {
            this.componentRef.setInput(key, value);
        }
    }
    setOutputs(outputs) {
        this.#outputRegistry.unsubscribeAll();
        for (const prop in outputs) {
            this.setOutput(prop, outputs[prop]);
        }
    }
    setOutput(outputName, emit) {
        if (!this.#componentData.allowedOutputNames.includes(outputName))
            return;
        if (!emit) {
            this.#outputRegistry.unsubscribe(outputName);
            return;
        }
        const hasListener = this.#outputRegistry.hasListener(outputName);
        this.#outputRegistry.setListener(outputName, emit);
        if (hasListener) {
            return;
        }
        const instance = this.componentRef.instance;
        const output = instance[outputName];
        if (output && output instanceof OutputEmitterRef) {
            output.subscribe(value => {
                this.#outputRegistry.getListener(outputName)?.(value);
            });
        }
    }
}
class FlexRenderComponentOutputManager {
    #outputSubscribers = {};
    #outputListeners = {};
    #valueDiffer;
    constructor(keyValueDiffers, initialOutputs) {
        this.#valueDiffer = keyValueDiffers.find(initialOutputs).create();
        if (initialOutputs) {
            this.#valueDiffer.diff(initialOutputs);
        }
    }
    hasListener(outputName) {
        return outputName in this.#outputListeners;
    }
    setListener(outputName, callback) {
        this.#outputListeners[outputName] = callback;
    }
    getListener(outputName) {
        return this.#outputListeners[outputName];
    }
    unsubscribeAll() {
        for (const prop in this.#outputSubscribers) {
            this.unsubscribe(prop);
        }
    }
    unsubscribe(outputName) {
        if (outputName in this.#outputSubscribers) {
            this.#outputSubscribers[outputName]?.unsubscribe();
            delete this.#outputSubscribers[outputName];
            delete this.#outputListeners[outputName];
        }
    }
    diff(outputs) {
        return this.#valueDiffer.diff(outputs ?? {});
    }
}
