// @tanstack/angular-table/build/lib/esm2022/flex-render/flex-render-component.mjs
import { reflectComponentType, } from '@angular/core';
/**
 * Helper function to create a [@link FlexRenderComponent] instance, with better type-safety.
 *
 * - options object must be passed when the given component instance contains at least one required signal input.
 * - options/inputs is typed with the given component inputs
 * - options/outputs is typed with the given component outputs
 */
export function flexRenderComponent(component, ...options) {
    const { inputs, injector, outputs } = options?.[0] ?? {};
    return new FlexRenderComponent(component, inputs, injector, outputs);
}
/**
 * Wrapper class for a component that will be used as content for {@link FlexRenderDirective}
 *
 * Prefer {@link flexRenderComponent} helper for better type-safety
 */
export class FlexRenderComponent {
    component;
    inputs;
    injector;
    outputs;
    mirror;
    allowedInputNames = [];
    allowedOutputNames = [];
    constructor(component, inputs, injector, outputs) {
        this.component = component;
        this.inputs = inputs;
        this.injector = injector;
        this.outputs = outputs;
        const mirror = reflectComponentType(component);
        if (!mirror) {
            throw new Error(`[@tanstack-table/angular] The provided symbol is not a component`);
        }
        this.mirror = mirror;
        for (const input of this.mirror.inputs) {
            this.allowedInputNames.push(input.propName);
        }
        for (const output of this.mirror.outputs) {
            this.allowedOutputNames.push(output.propName);
        }
    }
}
