// @tanstack/angular-table/build/lib/esm2022/flex-render/view.mjs
import { TemplateRef, Type } from '@angular/core';
import { FlexRenderComponent } from './flex-render-component';
export function mapToFlexRenderTypedContent(content) {
    if (content === null || content === undefined) {
        return { kind: 'null' };
    }
    if (typeof content === 'string' || typeof content === 'number') {
        return { kind: 'primitive', content };
    }
    if (content instanceof FlexRenderComponent) {
        return { kind: 'flexRenderComponent', content };
    }
    else if (content instanceof TemplateRef) {
        return { kind: 'templateRef', content };
    }
    else if (content instanceof Type) {
        return { kind: 'component', content };
    }
    else {
        return { kind: 'primitive', content };
    }
}
export class FlexRenderView {
    view;
    #previousContent;
    #content;
    constructor(initialContent, view) {
        this.#content = initialContent;
        this.view = view;
    }
    get previousContent() {
        return this.#previousContent ?? { kind: 'null' };
    }
    get content() {
        return this.#content;
    }
    set content(content) {
        this.#previousContent = this.#content;
        this.#content = content;
    }
}
export class FlexRenderTemplateView extends FlexRenderView {
    constructor(initialContent, view) {
        super(initialContent, view);
    }
    updateProps(props) {
        this.view.markForCheck();
    }
    dirtyCheck() {
        // Basically a no-op. When the view is created via EmbeddedViewRef, we don't need to do any manual update
        // since this type of content has a proxy as a context, then every time the root component is checked for changes,
        // the property getter will be re-evaluated.
        //
        // If in a future we need to manually mark the view as dirty, just uncomment next line
        // this.view.markForCheck()
    }
    onDestroy(callback) {
        this.view.onDestroy(callback);
    }
}
export class FlexRenderComponentView extends FlexRenderView {
    constructor(initialContent, view) {
        super(initialContent, view);
    }
    updateProps(props) {
        switch (this.content.kind) {
            case 'component': {
                this.view.setInputs(props);
                break;
            }
            case 'flexRenderComponent': {
                // No-op. When FlexRenderFlags.PropsReferenceChanged is set,
                // FlexRenderComponent will be updated into `dirtyCheck`.
                break;
            }
        }
    }
    dirtyCheck() {
        switch (this.content.kind) {
            case 'component': {
                // Component context is currently valuated with the cell context. Since it's reference
                // shouldn't change, we force mark the component as dirty in order to re-evaluate function invocation in view.
                // NOTE: this should behave like having a component with ChangeDetectionStrategy.Default
                this.view.markAsDirty();
                break;
            }
            case 'flexRenderComponent': {
                // Given context instance will always have a different reference than the previous one,
                // so instead of recreating the entire view, we will only update the current view
                if (this.view.eqType(this.content.content)) {
                    this.view.update(this.content.content);
                }
                this.view.markAsDirty();
                break;
            }
        }
    }
    onDestroy(callback) {
        this.view.componentRef.onDestroy(callback);
    }
}
